### PR TITLE
Remove Tolerance Check in yIndexOfX for Histogram Data

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR fixes a problem with the `yIndexOfX` method when used on histogram data. It does not make sense to have a tolerance for histogram data. Here's an example of a histogram with two bins (top row is bin centre values, bottom row is bin edges):
```
| 3.5 | 4.5 |
3     4      5
```
In this case of yIndexOfX(4.1) we should expect to get the index of the right hand bin back regardless of any tolerance. At the moment if you supply a tolerance value (eg 0.2) the function would return the left hand bin. This does not make sense.

**To test:**
Make sure the tests pass, and that the above example returns the expected bin index.

No related issue. *No release notes required as this is more of an internal change.*

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by DannyHindson — CHANGES_REQUESTED** [review]
Thanks for looking at this. I've tested it with this script and it now gives the correct answer (y index of 1 which is the second bin):
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws=CreateWorkspace(DataX=[3,4,5],DataY=[1,2])
xval=4.1
testx=ws.yIndexOfX(xval,0,0.2)
print("y index of " + str(xval) + "=" + str(testx))
```
I've added a couple of detail level comments - only non-trivial one is about the code that handles x values being in descending order.

File: Framework/API/src/MatrixWorkspace.cpp [hunk 1]
```diff
@@ -1388,44 +1387,45 @@ std::size_t MatrixWorkspace::yIndexOfX(const double xValue,
...
     // If we are pointing at the first value then we want to be in the first bin
     if (lowerIter == xValues.cbegin())
       ++lowerIter;
 
     hops = std::distance(xValues.cbegin(), lowerIter);
   } else {
-    auto lowerIter = std::upper_bound(xValues.crbegin(), xValues.crend(),
-                                      xValue + tolerance);
+    auto lowerIter =
+        std::upper_bound(xValues.crbegin(), xValues.crend(), xValue);
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply Framework/API/src/MatrixWorkspace.cpp]: Good spot, this is wrong 👍

File: Framework/API/inc/MantidAPI/MatrixWorkspace.h [hunk 2]
```diff
@@ -147,7 +147,7 @@ class MANTID_API_DLL MatrixWorkspace : public IMDWorkspace,
   Types::Core::DateAndTime getLastPulseTime() const;
 
   /// Returns the y index which corresponds to the X Value provided
-  std::size_t yIndexOfX(const double xValue, const std::size_t = 0,
+  std::size_t yIndexOfX(const double xValue, const std::size_t &index = 0,
```
Comment (DannyHindson) [inline Framework/API/inc/MantidAPI/MatrixWorkspace.h:150]: One more question on this - why did you change the index parameter here to be passed by reference? In various other places you changed some other parameters that had primitive types to be passed by value. That seemed sensible since I don't think there's any upside to passing them by const ref. But this one went in the other direction.

Comment (robertapplin) [reply Framework/API/inc/MantidAPI/MatrixWorkspace.h:150]: Your right, I don't think there is any benefit to passing `std::size_t` by reference as it is a primitive type. However, I like the consistency of passing all standard library data types (e.g. `std::string` , `std::vector` , `std::size_t`) by reference. From a stylistic viewpoint it just seems better (maybe only to me 😄)

Comment (DannyHindson) [reply Framework/API/inc/MantidAPI/MatrixWorkspace.h:150]: OK, seems like you're not alone looking elsewhere in the Mantid code eg
```
NumericAxis::NumericAxis(const std::size_t &length)
```
Fine to go with this

**Review by DannyHindson — APPROVED** [review]
